### ISSUE TITLE
kv: span use-after-finish race in RangeLookup->sendToReplicas

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#18538]
seen on release-1.1 in `TestReplicateReAddAfterDown`.

https://teamcity.cockroachdb.com/viewLog.html?buildId=353695&buildTypeId=Cockroach_UnitTests

```
==================
Write at 0x00c420580228 by goroutine 794:
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).reset()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:721 +0x182
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.freeTrace()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:879 +0x57
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).unref()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:838 +0x9b
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*traceBucket).Add()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:578 +0x133
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).Finish()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:386 +0x41f
  github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).FinishWithOptions()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:262 +0x192
  github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).Finish()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:246 +0x63
  github.com/cockroachdb/cockroach/pkg/util/tracing.FinishSpan()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:544 +0x52
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:270 +0xf4

Previous read at 0x00c420580228 by goroutine 236:
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).addEvent()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:767 +0x1b0
  github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).LazyPrintf()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:801 +0xdd
  github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).LogFields()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:340 +0xce2
  github.com/cockroachdb/cockroach/pkg/util/log.eventInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:131 +0x2ca
  github.com/cockroachdb/cockroach/pkg/util/log.ErrEventf()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:179 +0x7e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1171 +0xcad
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:382 +0x352
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:299 +0x301
  github.com/cockroachdb/cockroach/pkg/storage_test.mtcRangeDescriptorDB.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:636 +0xc1
  github.com/cockroachdb/cockroach/pkg/storage_test.(*mtcRangeDescriptorDB).RangeLookup()
      <autogenerated>:77 +0xf2
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:418 +0x4de
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:284 +0x11e
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/singleflight/singleflight.go:93 +0x4c
```

**Comment by tschottdorf** [comment]
This might be test-only but isn't necessarily. The race is the `ErrEventf` below:

```go
		case call := <-done:
			if err := call.Err; err != nil {
				// [omitted]
				if haveCommit && !grpcutil.RequestDidNotStart(err) {
					ambiguousError = err
				}
				log.ErrEventf(ctx, "RPC error: %s", err) // <- race here
```

while we're closing the span (at the end of a `RunAsyncTask`). Creating goroutine is below. My conjecture is that the call to `([]*Stores)[storeIdx].Send(ctx, ba)` can leak `ctx` to a `RangeLookup`-goroutine that ends up outliving the surrounding task. As such, it's possible that we also have this race in real code with the local-server optimization. But I haven't traced this down exactly.

```
Goroutine 794 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:270 +0x14c
  github.com/cockroachdb/cockroach/pkg/storage_test.(*multiTestContextKVTransport).SendNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:509 +0x50c
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1135 +0x6bc
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:382 +0x352
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:299 +0x301
  github.com/cockroachdb/cockroach/pkg/storage_test.mtcRangeDescriptorDB.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/client_test.go:636 +0xc1
  github.com/cockroachdb/cockroach/pkg/storage_test.(*mtcRangeDescriptorDB).RangeLookup()
      <autogenerated>:77 +0xf2
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:418 +0x4de
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:284 +0x11e
  github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/singleflight/singleflight.go:93 +0x4c
```

**Comment by andreimatei** [comment]
Seen again with the same test: https://teamcity.cockroachdb.com/viewLog.html?buildId=354804&buildTypeId=Cockroach_UnitTests_Testrace

**Comment by arjunravinarayan** [comment]
Bumping this issue again, and assigning to @bdarnell for triage.

**Comment by bdarnell** [comment]
I think this was fixed by #18611.
